ecc/bls12-377/fr/iop/utils.go

```diff
@@ -0,0 +1,378 @@
+// Copyright 2020 ConsenSys Software Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Code generated by consensys/gnark-crypto DO NOT EDIT
+
+package iop
+
+import (
+	"fmt"
+
+	"github.com/consensys/gnark-crypto/ecc/bls12-377/fr"
+	"github.com/consensys/gnark-crypto/ecc/bls12-377/fr/fft"
+)
+
+func printVector(v []fr.Element) {
+	fmt.Printf("[")
+	for i := 0; i < len(v); i++ {
+		fmt.Printf("Fr(%s), ", v[i].String())
+	}
+	fmt.Printf("]\n")
+}
+
+func printPolynomials(p []*Polynomial) {
+	fmt.Printf("[\n")
+	for i := 0; i < len(p); i++ {
+		printVector(p[i].Coefficients)
+		fmt.Printf(",\n")
+	}
+	fmt.Printf("]\n")
+}
+
+func printLayout(f Form) {
+
+	if f.Basis == Canonical {
+		fmt.Printf("CANONICAL")
+	} else if f.Basis == LagrangeCoset {
+		fmt.Printf("LAGRANGE_COSET")
+	} else {
+		fmt.Printf("LAGRANGE")
+	}
+	fmt.Println("")
+
+	if f.Layout == Regular {
+		fmt.Printf("REGULAR")
+	} else {
+		fmt.Printf("BIT REVERSED")
+	}
+	fmt.Println("")
+
+	if f.Status == Locked {
+		fmt.Printf("LOCKED")
+	} else {
+		fmt.Printf("UNLOCKED")
+	}
+	fmt.Println("")
+}
+
+type modifier func(p *Polynomial, d *fft.Domain) *Polynomial
+
+// return a copy of p
+func copyPoly(p Polynomial) Polynomial {
+	size := len(p.Coefficients)
+	var r Polynomial
+	r.Coefficients = make([]fr.Element, size)
+	copy(r.Coefficients, p.Coefficients)
+	r.Info = p.Info
+	return r
+}
+
+// return an ID corresponding to the polynomial extra data
+func getShapeID(p Polynomial) int {
+	return int(p.Info.Basis)*4 + int(p.Info.Layout)*2 + int(p.Info.Status)
+}
+
+//----------------------------------------------------
```

Comment:
mmhhh I don't like that part it doesn't seem really robust to changes; 
I think the combinatorial is not too crazy and this can be factored in a single method; 
```
isLocked := p.IsLocked()
if isLocked {
    p = p.Clone()
}
if targetBasis != p.Basis {
    // ...
}
```

at the minimum, would be nice to have some compiler-enforced consistency checks / safety.